src/Qir/Runtime/lib/QIR/QubitManagerRestrictedReuse.hpp

```diff
@@ -0,0 +1,210 @@
+    // This class maintains status of elements in the list by virtue of linking them as part of this list.
+    // This class doesn't update status of elementes excluded from the list.
+    // This class is small, contains no pointers and relies on default shallow copying/destruction.
+    struct QubitListInSharedArray final
+    {
+    private:
+        int firstElement = NoneMarker;
+        int lastElement = NoneMarker;
+
+    public:
+        // Initialize empty list
+        QubitListInSharedArray() = default;
+
+        // Initialize as a list with sequential elements from minElement to maxElement inclusve.
+        QubitListInSharedArray(int minElement, int maxElement, int* sharedQubitStatusArray);
+
+        bool IsEmpty();
+        void AddQubit(int id, bool addToFront, int* sharedQubitStatusArray);
+        int TakeQubitFromFront(int* sharedQubitStatusArray);
+        void MoveAllQubitsFrom(QubitListInSharedArray& source, int* sharedQubitStatusArray);
+    };
+
+
+    // Restricted reuse area consists of multiple segments. Qubits released in one segment cannot be reused in another.
+    // One restricted reuse area can be nested in a segment of another restricted reuse area.
+    // This class tracks current segment of an area. Previous segments are tracked collectively (not individually).
+    // This class is small, contains no pointers and relies on default shallow copying/destruction.
+    struct RestrictedReuseArea final
+    {
+    public:
+        QubitListInSharedArray FreeQubitsReuseProhibited;
+        QubitListInSharedArray FreeQubitsReuseAllowed;
+
+        RestrictedReuseArea() = default;
+        RestrictedReuseArea(QubitListInSharedArray freeQubits);
+    };
+
+
+
+    // This is NOT a stack! We modify it only by push/pop, but we also iterate over elements.
+    // TODO: Better name?
+    class CRestrictedReuseAreaStack final : public std::vector<RestrictedReuseArea>
+    {
+    public:
+        // No complex scenarios for now. Don't need to support copying/moving.
+        CRestrictedReuseAreaStack() = default;
+        CRestrictedReuseAreaStack(const CRestrictedReuseAreaStack&) = delete;
+        CRestrictedReuseAreaStack& operator = (const CRestrictedReuseAreaStack&) = delete;
+        ~CRestrictedReuseAreaStack() = default;
+
+        void PushToBack(RestrictedReuseArea area);
+        RestrictedReuseArea PopFromBack();
+        RestrictedReuseArea& PeekBack();
+        // TODO: Remove and use size directly?
+        int Count();
+    };
+
+    class CQubitManagerRestrictedReuse : public IQubitManager
+    {
+        // Restricted reuse area control, allocation and release
+
+        public: void StartRestrictedReuseArea();
+        public: void NextRestrictedReuseSegment();
+        public: void EndRestrictedReuseArea();
+        private: long AllocateQubit(); // Computation complexity is O(number of nested restricted reuse areas).
+        private: void ReleaseQubit(long id);
```

Comment:
> **long** AllocateQubit()

FYI for the future: I'm not sure if the qubit manager will be used _in addition_ to the [`FullStateSimulator`](https://github.com/microsoft/qsharp-runtime/blob/3f7ea4b7f0174b58f665737452da64710eb227c3/src/Qir/Runtime/lib/Simulators/FullstateSimulator.cpp#L193) wrapper class and [actual simulator](https://github.com/microsoft/qsharp-runtime/blob/3f7ea4b7f0174b58f665737452da64710eb227c3/src/Simulation/Native/src/simulator/capi.hpp#L72), or _instead_ of them, but eventually I believe we will likely use the same type everywhere to designate the qubit ID.

You may want to reuse one of the types they currently use for the qubit ID.

Reply:
The qubit manager will be in addition to the wrapper. I will replace the simple ++ logic of qubit allocation. Yes, it would be beneficial to designate a separate type for qubit id. I added QubitIdType. I made it signed. See separate discussion about signed/unsigned.